### PR TITLE
Fix #225: Make possible to call OnboardingService from Enrollment-Server

### LINKED ISSUE #225: Make possible to call OnboardingService from Enrollment-Server
After split of `enrollment-server` to two wars, it is not possible anymore to call `OnboardingService` place in `enrollment-server-onboarding` from `enrollment-server`. Extract minimal functionality to a separate module.

### PR DESCRIPTION
**Pull request by banterCZ**

### CODE REVIEW

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingServiceImpl.java [hunk 1]
```diff
@@ -47,20 +47,19 @@
...
 
 /**
- * Service implementing the onboarding process.
+ * Service implementing specific behavior for the onboarding process. Shared behavior is inherited from {@link CommonOnboardingService}.
  *
  * @author Roman Strobl, roman.strobl@wultra.com
  */
 @Service
-public class OnboardingService {
+public class OnboardingServiceImpl extends CommonOnboardingService {
```
Comment (banterCZ) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingServiceImpl.java:55]: We may change inheritance to composition, if wanted.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingServiceImpl.java [hunk 2]
```diff
@@ -73,15 +72,15 @@ public class OnboardingService {
      * @param otpService OTP service.
      */
     @Autowired
-    public OnboardingService(
+    public OnboardingServiceImpl(
```
Comment (banterCZ) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingServiceImpl.java:75]: Any better idea for name than `Impl`? Right now, we have `OnboardingServiceImpl` << `CommonOnboardingService` << `OnboardingService`

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OtpServiceImpl.java [hunk 3]
```diff
@@ -40,19 +38,17 @@
...
 /**
- * Service implementing OTP delivery and verification during onboarding process.
+ * Service implementing OTP delivery and verification specific for the onboarding process.
+ * Shared behavior is inherited from {@link CommonOtpService}.
  *
  * @author Roman Strobl, roman.strobl@wultra.com
  */
 @Service
-public class OtpService {
+public class OtpServiceImpl extends CommonOtpService {
```
Comment (banterCZ) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OtpServiceImpl.java:47]: dtto

File: enrollment-server/pom.xml [hunk 4]
```diff
@@ -43,10 +43,18 @@
...
+        <dependency>
+            <groupId>com.wultra.security</groupId>
+            <artifactId>enrollment-server-common</artifactId>
+        </dependency>
+
+        <!-- TODO (racansky, 2022-06-21) not sufficient isolation, enrollment shares some logic from onboarding but including all classes -->
         <dependency>
             <groupId>com.wultra.security</groupId>
-            <artifactId>enrollment-server-api-model</artifactId>
+            <artifactId>enrollment-server-onboarding</artifactId>
```
Comment (banterCZ) [inline enrollment-server/pom.xml:55]: I am not quite happy with this solution, but might be good enough as the very fist step.

File: enrollment-server/src/main/java/com/wultra/app/enrollmentserver/onboarding/activation/ActivationProcessService.java [hunk 5]
```diff
@@ -15,11 +15,11 @@
  * You should have received a copy of the GNU Affero General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
-package com.wultra.app.onboardingserver.activation;
+package com.wultra.app.enrollmentserver.onboarding.activation;
```
Comment (banterCZ) [inline enrollment-server/src/main/java/com/wultra/app/enrollmentserver/onboarding/activation/ActivationProcessService.java:18]: Reverted to `enrollmentserver`, should never been moved to `onboardingserver`.

File: enrollment-server/src/main/java/com/wultra/app/enrollmentserver/onboarding/activation/ActivationOtpService.java [hunk 6]
```diff
@@ -15,11 +15,11 @@
  * You should have received a copy of the GNU Affero General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
-package com.wultra.app.onboardingserver.activation;
+package com.wultra.app.enrollmentserver.onboarding.activation;
```
Comment (banterCZ) [inline enrollment-server/src/main/java/com/wultra/app/enrollmentserver/onboarding/activation/ActivationOtpService.java:18]: dtto

**Comment by banterCZ** [comment]
- [ ] test whether the downstream project work with this layout

File: enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/UpdateProcessRequest.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.Date;
+
+/**
+ * Request for {@link OnboardingService#updateProcess(UpdateProcessRequest)}.
+ *
+ * @author Lubos Racansky, lubos.racansky@wultra.com
+ */
+@Builder
+@Getter
+public final class UpdateProcessRequest {
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/UpdateProcessRequest.java:36]: Ok, perhaps we can move this class into a `model` package to separate it from others?

File: enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/OnboardingProcessException.java [hunk 8]
```diff
@@ -15,7 +15,7 @@
  * You should have received a copy of the GNU Affero General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
-package com.wultra.app.onboardingserver.errorhandling;
+package com.wultra.app.enrollmentserver.common.onboarding;
```
Comment (romanstrobl) [inline enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/OnboardingProcessException.java:18]: Can we keep this class in `errorhandling` package?

File: enrollment-server-onboarding/pom.xml [hunk 9]
```diff
@@ -164,6 +160,13 @@
 
     <build>
         <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-war-plugin</artifactId>
+                <configuration>
+                    <attachClasses>true</attachClasses>
```
Comment (romanstrobl) [inline enrollment-server-onboarding/pom.xml:167]: We do not use this option in other PowerAuth pom.xml files, is it necessary?

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingServiceImpl.java (hunk 2 above)
Comment (romanstrobl) [reply enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingServiceImpl.java:77]: I think it is ok.

File: enrollment-server/pom.xml (hunk 4 above)
Comment (romanstrobl) [reply enrollment-server/pom.xml:60]: We can improve the separation later.

File: enrollment-server/src/main/java/com/wultra/app/enrollmentserver/onboarding/activation/OnboardingComponentsConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+})
+@EntityScan(basePackages = {
+        "com.wultra.app.enrollmentserver.database.entity", // not to override component scan for enrollment-server
+        "com.wultra.app.onboardingserver.database.entity" // dependencies from enrollment-server-onboarding
+})
+@Configuration
+// TODO (racansky, 2022-06-21) improve isolation, enrollment includes all classes from onboarding
+public class OnboardingComponentsConfiguration {
+
+    @Bean
```
Comment (romanstrobl) [inline enrollment-server/src/main/java/com/wultra/app/enrollmentserver/onboarding/activation/OnboardingComponentsConfiguration.java:49]: Can we add JavaDoc for methods in this class?

File: enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/OnboardingService.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+import com.wultra.app.enrollmentserver.common.annotation.PublicApi;
+
+/**
+ * Service implementing the onboarding process.
+ *
+ * @author Lubos Racansky, lubos.racansky@wultra.com
+ */
+@PublicApi
+public interface OnboardingService {
```
Comment (romanstrobl) [inline enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/OnboardingService.java:28]: I suggest to put this class into `api` package.

File: enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/OtpService.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+import com.wultra.app.enrollmentserver.common.annotation.PublicApi;
+import com.wultra.app.enrollmentserver.model.enumeration.OtpType;
+
+/**
+ * Service implementing OTP verification during onboarding process.
+ *
+ * @author Lubos Racansky, lubos.racansky@wultra.com
+ */
+@PublicApi
+public interface OtpService {
```
Comment (romanstrobl) [inline enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/OtpService.java:30]: I suggest to put this class into `api` package.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/CommonOtpService.java [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+        OtpVerifyResponse response = new OtpVerifyResponse();
+        response.setProcessId(processId);
+        response.setOnboardingStatus(process.getStatus());
+        response.setExpired(expired);
+        response.setVerified(verified);
+        response.setRemainingAttempts(maxFailedAttempts - failedAttempts);
+        return response;
+    }
+
+    protected OnboardingProcessEntity failProcess(OnboardingProcessEntity entity, String errorDetail) {
```
Comment (romanstrobl) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/CommonOtpService.java:134]: Missing JavaDoc.

**Review by romanstrobl — APPROVED** [review]
In general, I think this is a step in the right direction, although we are not there yet with cleaning all the internal dependencies. I left some minor ideas for improvements and comments.

File: enrollment-server-onboarding/pom.xml (hunk 9 above)
Comment (banterCZ) [reply enrollment-server-onboarding/pom.xml:167]: 🤔 Interesting. Where do you use the war artifact as a dependency? This is according the documentation https://maven.apache.org/plugins/maven-war-plugin/war-mojo.html#attachClasses

It creates:
- `enrollment-server-onboarding-1.3.0.war`
- `enrollment-server-onboarding-1.3.0-classes.jar`

Which is used later on in 
```xml
<dependency>
    <groupId>com.wultra.security</groupId>
    <artifactId>enrollment-server-onboarding</artifactId>
    <version>${project.version}</version>
    <classifier>classes</classifier>
</dependency>
```

Maven does not build it without that but it is fine for IDE. Don't you have jar plugin even for war packaging?

File: enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/UpdateProcessRequest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (banterCZ) [reply enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/UpdateProcessRequest.java:36]: If there is no strict requirement, I would not over-engineer with packages. `UpdateProcessRequest` is related to `OnboardingService`, that's why it is placed in the same package so far.

File: enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/OnboardingProcessException.java (hunk 8 above)
Comment (banterCZ) [reply enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/OnboardingProcessException.java:18]: Yes, of course, I have been hesitating about that. I will move it to `com.wultra.app.enrollmentserver.common.onboarding.errorhandling`

File: enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/UpdateProcessRequest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (romanstrobl) [reply enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/UpdateProcessRequest.java]: We always put `model` classes in `model` package in PowerAuth, so I don't want to break this rule.

File: enrollment-server-onboarding/pom.xml (hunk 9 above)
Comment (romanstrobl) [reply enrollment-server-onboarding/pom.xml:167]: Ah, I see now. We always use jars for library dependencies, so we don't have an issue in other projects. I guess if the separation would be perfect, we would not need to use this option.

File: enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/UpdateProcessRequest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (banterCZ) [reply enrollment-server-common/src/main/java/com/wultra/app/enrollmentserver/common/onboarding/UpdateProcessRequest.java]: - [ ] Move `UpdateProcessRequest` to `model` package according to the project convention

**Pull request merged**